### DISCUSSION TITLE
pyyaml has been disabled because it does not meet homebrew/core's requirements for Python library formulae!

### DISCUSSION THREAD

**Discussion by stefanottili** [Homebrew/discussions#5707] [Everyday usage]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.4.3
ORIGIN: https://github.com/Homebrew/brew
HEAD: 91f3671b16fe62121364c465c14e0dfd088d7ead
Last commit: 6 days ago
Core tap JSON: 02 Nov 18:17 UTC
Core cask tap JSON: 02 Nov 18:17 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: :0
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 3.3.5 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.3.5/bin/ruby
CPU: octa-core 64-bit arm_firestorm_icestorm
Clang: 16.0.0 build 1600
Git: 2.47.0 => /opt/homebrew/bin/git
Curl: 8.7.1 => /usr/bin/curl
macOS: 15.1-arm64
CLT: 16.1.0.0.1.1729049160
Xcode: N/A
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Warning: Some installed kegs have no formulae!
This means they were either deleted or installed manually.
You should find replacements for the following formulae:
  cudd
  swig@3

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  swig@3
```

### Description of issue

https://theopenroadproject.org uses yaml and this worked til recently.
How do I get this to work again ?
```
Traceback (most recent call last):
  File "OpenROAD-flow-scripts/flow/scripts/defaults.py", line 4, in <module>
    import yaml
ModuleNotFoundError: No module named 'yaml'

brew install pyyaml
==> Downloading https://formulae.brew.sh/api/formula.jws.json
########################################################################################################################################################################### 100.0%
==> Downloading https://formulae.brew.sh/api/cask.jws.json
########################################################################################################################################################################### 100.0%
Error: pyyaml has been disabled because it does not meet homebrew/core's requirements for Python library formulae! It was disabled on 2024-10-06.
```

**Comment by iMichka** [comment]
Yes. You should use `pip3 install` for that package. We do not provide it anymore. There was no added value for homebrew to ship it when it can be early be installed though pip.

**Comment by stefanottili** [comment]
I'm sorry, I'm a complete python newbie, pyyaml was installed using the OpenRoad DependencyInstaller.sh

Using pip3 install gives me "error: externally-managed-environment".
Any hint on how to proceed ?
```
pip3 install pyyaml

[notice] A new release of pip is available: 24.2 -> 24.3.1
[notice] To update, run: python3.13 -m pip install --upgrade pip
error: externally-managed-environment

× This environment is externally managed
╰─> To install Python packages system-wide, try brew install
    xyz, where xyz is the package you are trying to
    install.
...
python3.13 -m pip install --upgrade pip

[notice] A new release of pip is available: 24.2 -> 24.3.1
[notice] To update, run: python3.13 -m pip install --upgrade pip
error: externally-managed-environment

× This environment is externally managed
╰─> To install Python packages system-wide, try brew install
    xyz, where xyz is the package you are trying to
    install.
...
```

**Comment by stefanottili** [comment]
pip3 -V
pip 24.2 from /opt/homebrew/lib/python3.13/site-packages/pip (python 3.13)

**Comment by stefanottili** [comment]
It seems that all the sudden, just using brew is not enough any more.
Now I have to wrap my head around "virtual env". Ok, grr ..

https://discuss.python.org/t/on-macos-14-pip-install-throws-error-externally-managed-environment/50352/5

**Reply by iMichka** [reply]
Yes we introduced mandatory virtual envs lately, because a lot of people try to install various versions of libraries and mixing stuff, and sometimes breaking their env. Virtual envs allow a clear separation between different envs and gives better control of what you are doing.

It's a little bit harder to setup the first time, but it's really better in the long run.

**Reply by kopaka7** [reply]
I find the virtual envs to be frustrating because I am having trouble installing system-wide packages on my Mac and I keep running into the issue where installs from brew are not being detected or now this disabled issue. I am trying to get pytorch to work on my Mac so I can use whisper.cpp. But despite running brew install pytorch, when I open Python in terminal, I can't import torch. I tried reinstalling and relinking it, but nothing. So I decided to build pytorch locally. But when I try to do this, I'm missing all sorts of packages in Python. And now brew is preventing me from installing them. Building pytorch locally and then installing it in a venv doesn't help me at all when I'm trying to get it working at the system level.

**Reply by SMillerDev** [reply]
If you need it for whisper.cpp, why not install that through brew?

**Comment by remkolems** [comment]
@stefanottili I ran into this as well. For beginners see my complete example below to get a head start.

---
1. Go to a directory (change directory directive). E.g.:
```shell
cd ~/siem-k8s-cluster/kubernetes/siem/apps/security/tetragon/app/tmp
```

2. Create a virtual environment (`api2jsonschema` can be any name, is the name of the folder that will be created)
```shell
python3 -m venv api2jsonschema
```

3. Use/load the newly created venv (`api2jsonschema/bin/activate` is path to file `activate`)
```shell
source api2jsonschema/bin/activate
```

4. Install a library
```shell
pip3 install PyYAML
```

5. Feedback of library install within venv
```shell
Collecting PyYAML
  Using cached PyYAML-6.0.2-cp313-cp313-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (2.1 kB)
Using cached PyYAML-6.0.2-cp313-cp313-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (759 kB)
Installing collected packages: PyYAML
Successfully installed PyYAML-6.0.2

[notice] A new release of pip is available: 24.2 -> 24.3.1
[notice] To update, run: pip install --upgrade pip
```

6. Upgrade pip within venv
```shell
pip3 install --upgrade pip
```

7. Feedback of the upgrade
```shell
Requirement already satisfied: pip in ./api2jsonschema/lib/python3.13/site-packages (24.2)
Collecting pip
  Using cached pip-24.3.1-py3-none-any.whl.metadata (3.7 kB)
Using cached pip-24.3.1-py3-none-any.whl (1.8 MB)
Installing collected packages: pip
  Attempting uninstall: pip
    Found existing installation: pip 24.2
    Uninstalling pip-24.2:
      Successfully uninstalled pip-24.2
Successfully installed pip-24.3.1
```

8. Do some actions with Python within the venv (`notice`: I'm still in the directory `~/siem-k8s-cluster/kubernetes/siem/apps/security/tetragon/app/tmp`. Any other directory is fine at this point. But I have not navigated to the subdirectory of the venv. Copying of files is not necessary).
```
> python3 openapi2jsonschema.py cilium.io_tracingpolicies.yaml
JSON schema written to tracingpolicy_v1alpha1.json
> python3 openapi2jsonschema.py cilium.io_podinfo.yaml
JSON schema written to podinfo_v1alpha1.json
> python3 openapi2jsonschema.py cilium.io_tracingpoliciesnamespaced.yaml
JSON schema written to tracingpolicynamespaced_v1alpha1.json
```

9. Exit or stop using the venv

```shell
deactivate
```

And visually what happens, when executing `python3 -m venv api2jsonschema` and `source api2jsonschema/bin/activate`

![image](https://github.com/user-attachments/assets/6ea75a4f-6736-4050-ae60-0d154b8c59cf)
![image](https://github.com/user-attachments/assets/5399e8e9-6a50-41d8-8504-efa788e91414)

**Comment by jghaines** [comment]
I had assumed that YAML had become part of python core. Now I'm forced to use a venv for a throwaway script that uses the YAML library. :-(

**Reply by amadeuspzs** [reply]
As per the output of the script, if you need this system-wide you can use

`pip3 install pyyaml --break-system-packages`
